### PR TITLE
mpu align for ro section of new memory region on non-XIP system

### PR DESCRIPTION
**Pull request by wentongwu**
1. On non-XIP system, SRAM is the default region, and relocated .data section and .bss section of SRAM shouldn't be inserted between _image_rom_start and _image_rom_end, because the memory region between _image_rom_start and _image_rom_end will construct the mpu read-only region. So this patch will put the relocated .data and .bss section of SRAM to mpu read-write region.

2. Also for the newly added memory region on non-XIP system, the relocated .text secition and .rodata section should also be mpu aligned. Add mpu_ro_region_start and mpu_ro_region_end for relocated section to help easily configure mpu.

3. And for code relocation on non-XIP system, the loading address shouldn't be FLASH, so this patch will distinguish the loading address for XIP and non-XIP system.

   Fixes: #16090.

### CODE REVIEW

**Comment by nashif** [comment]
Please update the title to what this PR is fixing, add a bug describing the bug you are fixing and then label the PR as Backport.

**Comment by nashif** [comment]
The body of the PR is very dense and difficult to read. Please restructure it a bit.

**Review by tyrantlizard77 — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -853,6 +853,10 @@ endif()
 
 if(CONFIG_CODE_DATA_RELOCATION)
    set(MEM_RELOCATAION_LD "${PROJECT_BINARY_DIR}/include/generated/linker_relocate.ld")
```
Comment (ioannisg) [inline CMakeLists.txt]: Isn't here a typo in the cmake variable?

Comment (wentongwu) [reply CMakeLists.txt]: > Isn't here a typo in the cmake variable?
yes, it should be RELOCATION. I will change it with next patch.

**Pull request merged**